### ISSUE TITLE
Most twisted.mail.pop3.IMailbox should be allowed to return Deferreds

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11128]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1701|
|Type|enhancement|
|Created|2006-05-04 20:22:03Z|

Requiring synchronous operation is unreasonable for large mailboxes.

<details><summary>Searchable metadata</summary>

```
trac-id__1701 1701
type__enhancement enhancement
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1146774123000000 1146774123000000
changetime__1147990689000000 1147990689000000
version__ 
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

I moved Cooperator from Epsilon into t.i and added support for returning a Deferred from listMessages.  Test coverage is also much improved and I think the protocol code is nicer now too.

Please review in async-pop3-mailbox-1701-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @radix|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @exarkun|
|-|-|

 * You might consider changing AsyncMailbox(DummyMailbox) to reactor.callLater(0) the calling back of its results, to catch any mistaken reliance on synchronous behavior.
 * LineBuffer could reasonably be renamed to something that doesn't imply newline-delimited data, given that it doesn't have any logic about dealing specifically with lines, just a sequence of writable objects. Just an observation.
 * You wrote a bunch of one-line comments above a bunch of small functions which are redundant and/or better off as docstrings.
 * Man, _longOperation is unfortunate. Not a blocker, but it kind of sucks that this code has to be in POP3.

```
             try:
                 i = int(i)
+                if i < 1:
+                    raise ValueError()
             except ValueError:
```

 * Man, that kind of makes me laugh/cry. Oh well.


```
+                    if err.check(ValueError, IndexError):
+                        # XXX TODO IndexError was supported for a while, but
+                        # really shouldn't be.  One error condition, one
+                        # exception type.  Deprecate this someday.
+                        self.failResponse("Invalid message-number: %r" % (i,))
```
 * How about adding a DeprecationWarning for the IndexError case?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @radix|
|-|-|

Changes made in rr16863, 16864, r16867, and r16868!  What say you?

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @exarkun|
|-|-|

Go ahead.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [16871]) Merge async-pop3-mailbox-1701-2

Author: exarkun
Reviewer: radix
Fixes [#11128](https://github.com/twisted/twisted/issues/11128)

This allows twisted.mail.pop3.IMailbox.listMessages to return Deferreds and updates the
protocol code to support this.  It also adds Cooperator to twisted.internet.task.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
